### PR TITLE
fix(sdk-core): Use PrebuiltTx with Tss Full SendMany

### PR DESCRIPTION
**Pull request by mohammadalfaiyazbitgo**
This pr enables rebuilding a prebuiltTx when used with SendMany on TSS Full wallets. Previously this method would not work because the SDK would be looking for a non-existent txRequestID from the prebuiltTx.

Ticket: WP-139

### CODE REVIEW

File: modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts [hunk 1]
```diff
@@ -369,4 +369,16 @@ export default class BaseTssUtils<KeyShare> extends MpcUtils implements ITssUtil
     const bitgoPublicKeyStr = response.publicKey as string;
     return readKey({ armoredKey: bitgoPublicKeyStr });
   }
+
+  /**
+   * Returns supported TxRequest versions for this wallet
+   */
+  public supportedTxRequestVersions(): TxRequestVersion[] {
```
[the remark this answers is not in the document]
Comment (zahin-mohammad) [reply modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts:376]: Tbh given these are tss utils, it should probablly happen at the consutrctor, but doesnt seem like we do that.

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 2]
```diff
@@ -1818,8 +1818,26 @@ export class Wallet implements IWallet {
...
+    if (
+      // verify the wallet must use txRequest Full api and must rebuild the tx before submitting
+      this._wallet.multisigType === 'tss' &&
+      !this.tssUtils?.supportedTxRequestVersions().includes('lite') &&
+      params.prebuildTx &&
+      !(typeof params.prebuildTx === 'string') &&
+      params.prebuildTx.buildParams &&
+      'preview' in params.prebuildTx.buildParams &&
+      params.prebuildTx.buildParams.preview
+    ) {
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:1831]: This condition is hard to read, please extract to a helper fn

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 3]
```diff
@@ -2990,7 +3008,7 @@ export class Wallet implements IWallet {
     }
 
     // ECDSA TSS uses TxRequestFull
-    if (this.baseCoin.getMPCAlgorithm() === 'ecdsa' || params.apiVersion === 'full' || this._wallet.type === 'cold') {
+    if (!this.tssUtils?.supportedTxRequestVersions().includes('lite') || params.apiVersion === 'full') {
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:3011]: Can we just define the apiVersion we are using at the top of this fn. would be easier to read/reuse.

**Review by zahin-mohammad — CHANGES_REQUESTED** [review]

File: modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts [hunk 4]
```diff
@@ -1353,6 +1362,40 @@ describe('TSS Ecdsa Utils:', async function () {
     });
   });
 
+  describe('supportedTxRequestVersions', function () {
+    it('returns only full for hot, cold, custodial wallet types', function () {
```
Comment (zahin-mohammad) [inline modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts:1366]: This should be separated into 3 separate tests.

File: modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts [hunk 5]
```diff
@@ -870,6 +875,83 @@ describe('TSS Utils:', async function () {
     });
   });
 
+  describe('supportedTxRequestVersions', function () {
+    it('should return full for custodial and cold wallets', async function () {
```
Comment (zahin-mohammad) [inline modules/bitgo/test/v2/unit/internal/tssUtils/eddsa.ts:879]: This should be two separate tests.

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 6]
```diff
@@ -284,6 +294,14 @@ export class Wallet implements IWallet {
     return this._wallet.coin;
   }
 
+  type(): WalletType | undefined {
+    return this._wallet.type;
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:298]: We can probably default to 'hot' when this value is undefined.

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 7]
```diff
@@ -1818,8 +1836,27 @@ export class Wallet implements IWallet {
       throw error;
     }
 
+    let txPrebuildQuery;
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:1839]: Is there a type we can assign here?

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 8]
```diff
@@ -1818,8 +1836,27 @@ export class Wallet implements IWallet {
...
+    const isTssWallet = this._wallet.multisigType === 'tss';
+    const supportedTxRequestVersions = this.tssUtils?.supportedTxRequestVersions() || [];
+    const mustUseTxRequestFull = supportedTxRequestVersions.length === 1 && supportedTxRequestVersions.includes('full');
+
+    if (
+      // verify the wallet must use txRequest Full api and must rebuild the tx before submitting
+      isTssWallet &&
+      mustUseTxRequestFull &&
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:1847]:
```
isTssWallet &&
mustUseTxRequestFull &&
```

seems redundant, 

can just use `mustUseTxRequestFull`

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 9]
```diff
@@ -2990,7 +3028,10 @@ export class Wallet implements IWallet {
     }
 
     // ECDSA TSS uses TxRequestFull
-    if (this.baseCoin.getMPCAlgorithm() === 'ecdsa' || params.apiVersion === 'full' || this._wallet.type === 'cold') {
+    const supportedTxRequestVersions = this.tssUtils?.supportedTxRequestVersions() ?? [];
+    const onlySupportsTxRequestFull =
+      supportedTxRequestVersions.length === 1 && supportedTxRequestVersions.includes('full');
+    if (apiVersion === 'full' || onlySupportsTxRequestFull) {
```
[the remark this answers is not in the document]
Comment (zahin-mohammad) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:3034]: We should throw an error in this case.

**Review by zahin-mohammad — CHANGES_REQUESTED** [review]

File: modules/sdk-core/src/bitgo/wallet/wallet.ts (hunk 6 above)
Comment (mohammadalfaiyazbitgo) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:298]: that's not gonna match the typing on the class.

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 10]
```diff
@@ -2990,7 +3028,10 @@ export class Wallet implements IWallet {
     }
 
     // ECDSA TSS uses TxRequestFull
-    if (this.baseCoin.getMPCAlgorithm() === 'ecdsa' || params.apiVersion === 'full' || this._wallet.type === 'cold') {
+    const supportedTxRequestVersions = this.tssUtils?.supportedTxRequestVersions() ?? [];
+    const onlySupportsTxRequestFull =
+      supportedTxRequestVersions.length === 1 && supportedTxRequestVersions.includes('full');
```
[the remark this answers is not in the document]
Comment (mohammadalfaiyazbitgo) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:3033]: https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types , compiler will complain if the var is being used in ways that don't make sense with the type.

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 11]
```diff
@@ -1818,8 +1836,27 @@ export class Wallet implements IWallet {
       throw error;
     }
 
+    let txPrebuildQuery;
+    const isTssWallet = this._wallet.multisigType === 'tss';
+    const supportedTxRequestVersions = this.tssUtils?.supportedTxRequestVersions() || [];
+    const mustUseTxRequestFull = supportedTxRequestVersions.length === 1 && supportedTxRequestVersions.includes('full');
```
[the remark this answers is not in the document]
Comment (mohammadalfaiyazbitgo) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:1842]: not magic strings, both 'tss' and 'full' are valid constrained types: https://betterprogramming.pub/understanding-string-literals-in-typescript-72862ecfca3e

File: modules/sdk-core/src/bitgo/wallet/wallet.ts (hunk 7 above)
Comment (mohammadalfaiyazbitgo) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:1839]: `Promise<PrebuildTransactionResult | string>`

File: modules/sdk-core/src/bitgo/wallet/wallet.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (mohammadalfaiyazbitgo) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:1842]: will follow up with a refactoring on the types

**Review by zahin-mohammad — APPROVED** [review]

**Pull request merged**